README.md

```diff
@@ -1,277 +1,899 @@
+The beacon information in the datum prevents misuse of beacons. The validator forces all beacons to
+be burned instead of being withdrawn. This ensures that beacons can only ever be located at
+swap addresses. **If the wrong beacon information is supplied, assets can be locked forever.** For
+this reason, all minting policies check that the datums have the proper information; the mint
+transaction will fail if they do not.
+
+The `Rational` type is a fraction (decimal types do not work on-chain). All prices in Cardano-Swaps
+are relative (similar to limit orders in an order-book exchange). Swaps are always priced in
+askedAsset/offeredAsset. For example, if ADA is being offered for DUST at a price of 1.5
+(converted to 3/2), the contract requires that 3 DUST are deposited for every 2 ADA removed from
+the swap address. Ratios of DUST:ADA >= 3/2 will pass, while ratios < 3/2 will fail. 
+
+When engaging in swaps, it is only necessary that the desired swap ratio is met; **not all assets
+in the UTxO must be swapped.** For example, if there is 100 ADA in a swap address requesting 2:1 for
+DUST, a user may swap 20 ADA, as long as they return 80 ADA and 10 DUST in the same Tx. Since every
+user explicitly defines their desired swap ratios, oracles are not required. The "global" price
+naturally emerges where the local bids and asks meet - just like an order-book. 
+
+**All prices for ADA must be in units of Lovelace.**
+
+##### Creating a One-Way Swap
+
+Swap "writers" must first create a swap by using the `CreateSwap` Redeemer in the following manner:
+
+In order to mint beacons with this redeemer, **all of the following must be true**:
+
+1) The beacons must go to an addres protected by the one-way swap validator script.
+2) The beacons must go to an address using a valid staking credential.
+3) The UTxOs with the beacons must have the proper value:
+    - Exactly two kinds of beacons: pair beacon and offer beacon.
+    - The beacons must correspond to the beacons in the datum.
+    - There must be exactly 1 of each beacon.
+    - No extraneous assets are in the UTxO. ADA is always allowed.
+4) The beacons must be stored with the proper inline `SwapDatum`:
+    - `beaconId` == this policy id.
+    - `pairBeacon` == pair beacon asset name for this swap.
+    - `offerId` == asset_id of the offer asset.
+    - `offerName` == asset_name of the offer asset.
+    - `offerBeacon` == offer beacon asset name for this swap's offer asset.
+    - `askId` == asset_id of the ask asset.
+    - `askName` == asset_name of the ask asset.
+    - `swapPrice` denominator > 0
+    - `swapPrice` > 0
+5) The offer asset and ask assets must be different assets.
```

Comment:
I am just trying to prompt your thinking here:  Do you need to add something about how the staking key is obtained?  For example, is there a check that the assets locked in the swap came from an address with the same staking key?  If any staking key can be specified does this open up the possibility for attacks by malicious web wallets and franken addresses somehow?

Reply:
When spending a swap UTxO, the staking credential is gotten directly from the UTxO itself. This information is passed to the script by the node itself and is unspoofable. So there is no risk from malicious users somehow providing an incorrect staking credential.

As far as where the assets go, I had originally required the assets go to the same address as the origin UTxO but, since Alice is explicitly approving the transaction, this seemed unnecessary. What if Alice was just moving her swap to a swap address with a different staking credential? Requiring the assets be locked at the same address as the origin means the swap would first have to be closed in one tx and then opened in the new address in another tx. I don't see any security issues with allowing the assets to go directly to another swap address. The requirement that the beacons must always go to a swap address is still properly enforced.

Reply:
So it would be impossible to create a swap from an address without staking key (ie: address only)?

I don't see a problem with that though.

Reply:
> So it would be impossible to create a swap from an address without staking key (ie: address only)?

It is possible to create a UTxO at this address but it is not possible to tag it with the beacons. Addresses cannot *deny* UTxO creation just like stake pool operators cannot deny delegation. 

Since there is no staking credential, there is effectively no "admin" for the address. I did not want to have any locked assets so I set this edge case to always be spendable by anyone. This address is effectively first come, first serve. I don't think this address should ever be used though which is why the beacons cannot go to an address without a staking credential.